java/dagger/internal/codegen/BindingGraph.java

```diff
@@ -1120,6 +1122,24 @@ private boolean hasLocallyPresentOptionalBinding(ResolvedBindings resolvedBindin
               && !getLocalExplicitBindings(keyFactory.unwrapOptional(resolvedBindings.key()).get())
                   .isEmpty();
         }
+
+        /**
+         * Returns {@code true} if {@code resolvedBindings} contains a provision binding
+         * for which there is an explicit present binding in this component that was resolved
+         * previously.
+         */
+        boolean hasLocallyConflictingContributionBinding(ResolvedBindings resolvedBindings) {
+          return resolvedBindings
+                .contributionBindings()
+                .stream()
+                .anyMatch(contributionBinding -> contributionBinding.bindingKind().equals(PROVISION)
+                  && !getLocalExplicitBindings(resolvedBindings.key()).isEmpty()
+                  && parentResolver.isPresent()
```

Comment:
We must check the parent here otherwise otherwise we could end up resolving a previously unresolved binding in the parent using a binding thats defined in the child:

```
@Component(modules = ParentModule.class) interface Parent {
  Child child();
}

@Subcomponent(modules = ChildModule.class) interface Child {
  Bar bar()
  Foo foo()
}

@Module abstract class ParentModule {
  @Provides static Foo foo(Bar bar) { ... }
}

@Module abstract class ChildModule {
  @Provides static Bar bar() { ... }
}
```